index.js

```diff
@@ -582,9 +582,92 @@ async function apostrophe(options, telemetry, rootSpan) {
     return synth;
   }
 
+  // Reorder modules based on their `before` property.
+  function sortModules(moduleNames) {
+    const definitions = self.options.modules;
+    // The module names that have a `before` property
+    const beforeModules = [];
+    // The metadata quick access of all modules
+    const modules = {};
+    // Recursion guard
+    const recursionGuard = {};
+    // The sorted modules result
+    const sorted = [];
+
+    // The base module sort metadata
+    for (const name of moduleNames) {
+      if (definitions[name].before) {
+        beforeModules.push(name);
+      }
+      modules[name] = {
+        before: definitions[name].before,
+        beforeSelf: []
+      };
+    }
+
+    // Loop through the modules that have a `before` property,
+    // validate and fill the initial `beforeSelf` metadata (first pass).
+    for (const name of beforeModules) {
+      const m = modules[name];
+      const before = m.before;
+      if (m.before === name) {
+        console.warn(`Module "${name}" has a 'before' property that references itself. Skipping.`);
+        continue;
+      }
+      if (!modules[before]) {
+        console.warn(`Module "${name}" has a 'before' property that references a non-existent module: "${before}". Skipping.`);
```

Comment:
Throw an error, stop the show. This is a new feature and we get to insist on good behavior.